### PR TITLE
Cob/namespaced structs

### PR DESCRIPTION
**Pull request by coburncoburn**
This patch adds struct namespacing, so structs from libararies will work.
This patch adds constructor, fallback, receive, and event abis to the Contract Struct, but does not use them in Geno
This patch parses numbers of 32 and less bytes correctly in Geno
This patch fixes a bug where arrays in function inputs would not generate the correct encoder in Geno.

### CODE REVIEW

File: Tests/GenoTests/Structs.swift [hunk 1]
```diff
@@ -131,4 +161,23 @@ enum Structs {
...
     }
+
+    static let lookAtRatFn = ABI.Function(
+        name: "lookAtRat",
+        inputs: [.tuple([.address]), .array(.address)],
+        outputs: [.address]
+    )
+
+    static func lookAtRat(rat: Animal.Rat, holes: [EthAddress]) throws -> EthAddress {
+        let query = try lookAtRatFn.encoded(with: [rat.asValue, .array(.address, holes.map { .address($0) })])
```
Comment (coburncoburn) [inline Tests/GenoTests/Structs.swift:172]: the way these array parameters were being generated was wrong

File: Sources/Geno/Geno.swift [hunk 2]
```diff
@@ -144,11 +144,23 @@ func outParameters(f: Contract.ABI.Function) -> String {
 }
 
 func callParameters(f: Contract.ABI.Function) -> String {
-    return f.inputs.map { parameter in
-        if structName(parameter) != nil {
-            "\(parameter.name).asValue"
+    return f.inputs.map { p in
```
Comment (coburncoburn) [inline Sources/Geno/Geno.swift:147]: this is depressingly close to `parameterToMatchableValueType` but unfortunately slightly different in output.

File: Sources/Geno/Geno.swift [hunk 3]
```diff
@@ -144,11 +144,23 @@ func outParameters(f: Contract.ABI.Function) -> String {
...
+            }
+        } else if isStruct(p) {
+            return "\(p.name).asValue"
+        } else if isTuple(p) {
+            let componentTypes = p.components!.enumerated().map { _, p in parameterToValueType(p, allowSchema: true) }
+            return ".tuple\(componentTypes.count)(\(componentTypes.joined(separator: ",\n ")))"
         } else {
-            "\(parameterToValueType(parameter, allowSchema: true))(\(parameter.name))"
+            // turning them into the enum values, with name values
+            return ".\(p.type)(\(p.name))"
```
Comment (coburncoburn) [inline Sources/Geno/Geno.swift:163]: this is depressingly close to `parameterToMatchableValueType` but unfortunately slightly different in output.

**Review by hayesgm — APPROVED** [review]
this looks _so good_

**Pull request merged**